IECoreAppleseed: Renderer::setOption changes and improvements

est77 opened the pull request:
- set decorrelate_pixels automatically depending on render passes.
- if the rendering_threads option is zero, use all CPU cores for rendering.
- if the as:cfg:pt:max_ray_intensity option is zero, disable max intensity clamping.
- disable max number of bounces for max_path_length options set to zero.
- check the type of the Data passed to setOption before accessing it.

johnhaddon commented:
Is there a case to be made for changing the default in Appleseed instead? I'm quite happy to merge this on your say-so, but it does seem odd to me that the renderer default wouldn't be something we want to honour.

est77 commented:
I moved the code to disable unlimited path lenghts to Gaffer. 
We still want to disable it and remove all the russian roulette related parameters in Gaffer, but users of IECoreAppleseed should not be affected by that.

I improved a few things and rebased the PR

johnhaddon commented on contrib/IECoreAppleseed/src/IECoreAppleseed/RendererImplementation.cpp:
```diff
@@ -170,16 +170,86 @@ void IECoreAppleseed::RendererImplementation::setOption( const string &name, Con
+						m_project->configurations().get_by_name( "final" )->get_parameters().remove_path( optName.c_str() );
+						m_project->configurations().get_by_name( "interactive" )->get_parameters().remove_path( optName.c_str() );
+						return;
+					}
```
Should there be an `else` here where we set the number of threads to the non-zero value?

est77 commented:
If the number of threads is non-zero the zero special case block is skipped and the number of threads (and the rest of the render settings) are set a bit later in the function, here:
https://github.com/appleseedhq/cortex/blob/IECoreAppleseed/contrib/IECoreAppleseed/src/IECoreAppleseed/RendererImplementation.cpp#L251

johnhaddon commented:
Ah yes - thanks for pointing that out, and sorry for not reading the code more closely in the first place.
